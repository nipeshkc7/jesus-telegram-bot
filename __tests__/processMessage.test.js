const { processMessage } = require('../Jesus/processMessage');

let people;
let peopleOwed;

describe('Testing message processing', () => {

    beforeEach(() => {
        people = {
            'John': {
                'spent': 0,
                'owes': {
                    'Mary': 0,
                    'Bob': 0,
                }
            },
            'Mary': {
                'spent': 0,
                'owes': {
                    'John': 0,
                    'Bob': 0,
                }
            },
            'Bob': {
                'spent': 0,
                'owes': {
                    'John': 0,
                    'Mary': 0,
                }
            },
        };

        peopleOwed = {
            'John': {
                'spent': 100,
                'owes': {
                    'Mary': 200,
                    'Bob': 100,
                }
            },
            'Mary': {
                'spent': 100,
                'owes': {
                    'John': 40,
                    'Bob': 30,
                }
            },
            'Bob': {
                'spent': 90,
                'owes': {
                    'John': 100,
                    'Mary': 80,
                }
            },
        };
    })

    test.each([['John', 300, 100], ['Mary', 600, 200], ['Bob', 100, 33.33]])(
        "should divide bills equally for %s", (user, spent, divided) => {
            const result = processMessage({
                from: {
                    first_name: user
                },
                text: `jesus i spent ${spent} dollars on food`
            }, people);

            Object.keys(result.peopleRecord).forEach(entry => {
                if (entry !== user) {
                    expect(result.peopleRecord[entry].owes[user]).toEqual(divided);
                }
            })
        });

    test.each([['John', 300, 100], ['Mary', 600, 200], ['Bob', 100, 33.33]])(
        "should divide bills equally for %s", (user, spent, divided) => {
            const result = processMessage({
                from: {
                    first_name: user
                },
                text: `jesus i spent ${spent} dollars on food`
            }, peopleOwed);

            expect(result.peopleRecord).toMatchSnapshot();
        });

    test.each(['John', 'Mary', 'Bob'])(
        "should clear bills for %s", (user) => {
            const result = processMessage({
                from: {
                    first_name: user
                },
                text: 'jesus i cleared my bills'
            }, people);
            expect(result.peopleRecord[user].owes).toEqual({});
        }
    )

    test.each(['John', 'Mary', 'Bob'])(
        "should return correct reply for %s", (user) => {
            const result = processMessage({
                from: {
                    first_name: user
                },
                text: 'jesus how much do i owe?'
            }, peopleOwed);
            expect(result.reply).toMatchSnapshot();
        }
    )

    it('should throw error if no user', () => {
        expect(() => {
            processMessage({ text: 'sth' }, {});
        }).toThrow();
    });
})