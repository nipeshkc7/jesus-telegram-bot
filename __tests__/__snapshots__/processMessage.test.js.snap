// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing message processing should divide bills equally for Bob 1`] = `
Object {
  "Bob": Object {
    "owes": Object {
      "John": 0,
      "Mary": 16.67,
    },
    "spent": 190,
  },
  "John": Object {
    "owes": Object {
      "Bob": 33.329999999999984,
      "Mary": 160,
    },
    "spent": 100,
  },
  "Mary": Object {
    "owes": Object {
      "Bob": 0,
      "John": 0,
    },
    "spent": 100,
  },
}
`;

exports[`Testing message processing should divide bills equally for John 1`] = `
Object {
  "Bob": Object {
    "owes": Object {
      "John": 100,
      "Mary": 50,
    },
    "spent": 90,
  },
  "John": Object {
    "owes": Object {
      "Bob": 0,
      "Mary": 60,
    },
    "spent": 400,
  },
  "Mary": Object {
    "owes": Object {
      "Bob": 0,
      "John": 0,
    },
    "spent": 100,
  },
}
`;

exports[`Testing message processing should divide bills equally for Mary 1`] = `
Object {
  "Bob": Object {
    "owes": Object {
      "John": 0,
      "Mary": 250,
    },
    "spent": 90,
  },
  "John": Object {
    "owes": Object {
      "Bob": 0,
      "Mary": 360,
    },
    "spent": 100,
  },
  "Mary": Object {
    "owes": Object {
      "Bob": 0,
      "John": 0,
    },
    "spent": 700,
  },
}
`;

exports[`Testing message processing should return correct reply for Bob 1`] = `"Bob owes:John 100 dollars,Mary 80 dollars,"`;

exports[`Testing message processing should return correct reply for John 1`] = `"John owes:Mary 200 dollars,Bob 100 dollars,"`;

exports[`Testing message processing should return correct reply for Mary 1`] = `"Mary owes:John 40 dollars,Bob 30 dollars,"`;
